plugin/documents.py

```diff
@@ -606,46 +614,70 @@ def render_highlights_on_main_thread() -> None:
 
     # --- textDocument/complete ----------------------------------------------------------------------------------------
 
-    def _on_query_completions_async(self, resolve: ResolveCompletionsFn, location: int) -> None:
-        session = self.session('completionProvider', location)
-        if not session:
-            resolve([], 0)
+    def _on_query_completions_async(self, resolve_completion_list: ResolveCompletionsFn, location: int) -> None:
+        sessions = self.sessions('completionProvider')
+        if not sessions:
+            resolve_completion_list([], 0)
             return
         self.purge_changes_async()
-        can_resolve_completion_items = bool(session.get_capability('completionProvider.resolveProvider'))
-        config_name = session.config.name
-        session.send_request_async(
-            Request.complete(text_document_position_params(self.view, location), self.view),
-            lambda res: self._on_complete_result(res, resolve, can_resolve_completion_items, config_name),
-            lambda res: self._on_complete_error(res, resolve))
-
-    def _on_complete_result(self, response: Optional[Union[dict, List]], resolve: ResolveCompletionsFn,
-                            can_resolve_completion_items: bool, session_name: str) -> None:
-        response_items = []  # type: List[Dict]
-        flags = 0
+
+        completion_promises = []  # type: List[Promise[CompletionResponseWithSession]]
+        for session in sessions:
+
+            def completion_request() -> Promise:
+                return Promise(lambda resolve: session.send_request_async(
+                    Request.complete(text_document_position_params(self.view, location), self.view),
+                    lambda res: self._on_complete_result(res, resolve, session),
+                    lambda res: self._on_complete_error(res, resolve, session)))
+
+            completion_promises.append(completion_request())
+
+        LspResolveDocsCommand.completions = []
+        Promise.all(completion_promises).then(
+            lambda responses: self._on_all_settled(responses, resolve_completion_list))
+
+    def _on_complete_result(self, response: CompletionResponse,
+                            resolve_promise: ResolveFunc[CompletionResponseWithSession], session: Session) -> None:
+        resolve_promise((response, session))
+
+    def _on_complete_error(self, error: dict,
+                           resolve_promise: ResolveFunc[CompletionResponseWithSession], session: Session) -> None:
+        resolve_promise((None, session))
+        sublime.status_message('Completion error: ' + str(error.get('message')))
+
+    def _on_all_settled(
+        self,
+        responses: Optional[List[CompletionResponseWithSession]],
+        resolve_completion_list: ResolveCompletionsFn
+    ) -> None:
+        if not responses:
+            resolve_completion_list([], 0)
+            return
```

Comment:
The generic type T in the `then` call is wrapped as `Optional[T]`
>    def then(self, onfullfilled: FullfillFunc**[Optional[T]**, TResult]) -> 'Promise[TResult]': 
 https://github.com/sublimelsp/LSP/blob/st4000-exploration/plugin/core/promise.py#L160

Maybe it is not necessary the type to be optional, and could be just `T`, 
>    def then(self, onfullfilled: FullfillFunc[**T**, TResult]) -> 'Promise[TResult]': 

If the type wasn't optional I could delete this if check above. (I have a feeling that the if check is not necessary, it will always be an array of resolved promise values)

@rchl do you think that we can remove the `Optional[T]`  type and just leave `T`? (I would maybe not do it in this PR, this was just something that I have noticed)

Reply:
Changing types in the `Promise` can have some unexpected cascading effects but I can try. :)